docs/spec_draft.md

```diff
@@ -1243,6 +1244,78 @@ Numeric precision is implementation-defined.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.slice
+
+### Semantics
+
+Extracts a sub-tensor from the `operand` using the dimension indices from the
+operands.
+
+More formally, `result[i0, ..., ik, ..., iR-1] = operand[j0+s0, ..., jk+sk, ..., jR-1+sR-1]`
+for all `sk` in `strides` for all `jk` in `start_indices[k]` $\leq$ `jk` $\lt$
+`limit_indices[k]` where 0 $\leq$ `k` $\lt$ rank(`operand`).
+
+Shape is calculated by `dim(result, i)` = $\lceil$`(limit_indices[i]-start_indices[i])/stride[i]`$\rceil$
+for `i`th dimension size of `operand`.
+
+### Operands
+
+| Name            | Type                                                                                                                                    |
+|-----------------|-----------------------------------------------------------------------------------------------------------------------------------------|
+| `operand`       | tensor of any supported types                                                                                                           |
+| `start_indices` | 1-dimensional array of integers containing the starting indices of the slice for each dimension                                         |
+| `limit_indices` | 1-dimensional array of integers containing the ending indices (exclusive) for the slice for each dimension                              |
+| `strides`       | 1-dimensional array of integers that decides the input stride of the slice. The slice picks every `strides[d]` element in `d` dimension |
+
+### Results
+
+| Name     | Type                          |
+|----------|-------------------------------|
+| `result` | tensor of any supported types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the same element type.
+  * (C2) `start_indices`, `limit_indices`, and `strides` have the same type.
+  * (C3) size(`start_indices`) = size(`limit_indices`) = size(`strides`) =
+  rank(`operand`).
+  * (C4) 0 $\le$ `start_indices[d]` $\le$ `limit_indices[d]` $\le$
```

Comment:
`limit_indices` are exclusive. So I guess when the start_indices == limit_indices, we might get a zero value as the dimension size.

Reply:
Yes, you are right, and this is allowed.

Reply:
You are right.
With zero dimension size, in any dimension, the number of elements a tensor can store is zero: will be an empty tensor.

Reply:
Correct, it would be an empty tensor. Though I think during implementation, we may do something along the lines of:
```
while (start_indices[i] < limit_indices[i] ...) {}
```
Since the indices must be less than or equal to size(`operand[i]`) (C4), the loop will terminate since `start_indices[i] == limit_indices[i] == 0` in which case there is no need to handle special cases of 0-dimension.

This is speculation though, so things might change once we actually start implementing this op.

Did you want me to add anything specific to the Semantics section to clarify empty tensors?

Reply:
Nice option!

>> Did you want me to add anything specific to the Semantics section to clarify empty tensors?
We do not have to write anything specific for 0 dimension size.

Reply:
"Though I think during implementation, we may do something along the lines of `while (start_indices[i] < limit_indices[i] ...) {}`". I don't think any explicit comparisons will be needed in the implementation of the interpreter, because the only place where these comparisons should be done is type inference (or constraints in the spec terminology). Also cf. my comment about the proposed formula for the semantics sections.

Reply:
You're right, I was viewing `start_indices` as the current index. My intent was to initialize `curr_index` as `start_indices` and iterating `curr_index` to select elements until it breaks the condition: `while (curr_index < limit_indices)`.

Reply:
I don't think that we need even that. Because our formula expresses the op as a transformation of index spaces, we can say something like (the code below uses operator overloads which don't exist at the moment):

```
Tensor eval(SliceOp op, const Tensor &operand) {
  Tensor result(op.getType());
  for (auto resultIt = result.index_begin(); resultIt != result.index_end(); ++resultIt) {
    auto operandIndex = op.getStartIndices() + *resultIt * op.getStrides();
    result.set(*resultIt, operand.get(operandIndex));
  }
  return result;
}
```

Note how `curr_index < limit_indices` never comes up in this loop. That's because we know that `op.getType()` is correct, so we know that `operandIndex` will always be in bounds. I think this shows the power of decoupling constraints from execution.